rust/kernel/src/sync.rs

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use core::cell::UnsafeCell;
+use core::fmt;
+use core::ops::{Deref, DerefMut};
+
+use crate::bindings;
+
+pub struct Mutex<T: ?Sized> {
+    lock: bindings::mutex,
+    data: UnsafeCell<T>,
+}
+
+impl<T> Mutex<T> {
+    /// Create a new Mutex
+    pub fn new(data: T, _name: &'static str) -> Self {
+        let lock = bindings::mutex::default();
+        // TODO: write mutex debug traces like .magic.
+        // TODO: use name in the debug version
+
+        Self {
+            data: UnsafeCell::new(data),
+            lock,
+        }
+    }
+}
+
+impl<T: ?Sized> Mutex<T> {
+    /// acquire a lock on the mutex
+    /// # unsafe
+    /// This is unsafe, as it returns a second lock if one is already help by the current process
```

Comment:
Is there any kernel APIs we can use to make this safe? It'd be a shame if all mutex usage requires `unsafe`.

Reply:
I guess we could test if the mutex is already locked and panic in that case (similar to what Rust's std does), but that would pretty much be like redoing the kernel mutex debugging infrastructure, so not sure how it will be received. Even if we go that route, we should have both APIs.

Reply:
^ "panic in that case" -> "deadlock in that case" -- considering we are the kernel, it is way better to deadlock a task (with perhaps a warning or even an oops prior to the deadlock itself) than kill the entire system.

---

Hmm... I wonder, could we do this statically for a subset of the use cases? Say, two types, `Mutex` and `MutexLocked` which move themselves into each other on every operation (& forget themselves to avoid calling unnecessary drops which would require a test). Then we could have code like:

	let m = m.lock();
	...
	let m = m.unlock();

Things like `trylock` could return an enum of either locked/unlocked, there could be a `m.take(|...| { ... })`, etc.

Of course, this doesn't work if one needs to keep the mutex stored in different states, but in some use cases this could be enough.

What do you think?

Reply:
Not a fan. With mutexguard you only need a reference of the mutex. Your idea needs a owned type which can not live in a struct or similar.

Reply:
I must be missing something here. Given that mutexes are not recursive, how is it that lock() can return a second MutexGuard when the mutex is already locked (by whomever)?

Reply:
If I understand it correctly, it tracks that we are the lock holder, and so mutex_lock does not block. This returns then a new lock.

Reply:
It does indeed track the owner, but it is not used to allow the same thread to recursively acquire the mutex; doing so will result in a deadlock.

This function doesn't need to be unsafe.

Reply:
Currently it's unsafe, as if you call Ut 2 times from the same function, it calles 2 times mutex_unlock, which results in a kernel panic.

Reply:
I tested this and it just deadlocked the task, perhaps we have different debugging configurations. In any case, whether it deadlocks or panics, why does it make this unsafe?

[Mutex](https://doc.rust-lang.org/std/sync/struct.Mutex.html) from the std library is not unsafe and behaves similarly:

> The exact behavior on locking a mutex in the thread which already holds the lock is left unspecified. However, this function will not return on the second call (it might panic or deadlock, for example).

Reply:
> why does it make this unsafe?

It depends whether the panic is a controlled one or one triggered due to a double free, GPF or similar.

> Mutex from the std library is not unsafe and behaves similarly:

Yeah, this is what I mentioned in my first comment above.

Reply:
We can remove them unsafe again, and Wirte some documentation. I can only do it after my exams. So either feel free to take my work, or wait until I have more time again.

Reply:
Don't worry about it. Good luck on your exams! :)